cartesian_trajectory_controller/test/test_controller.py

```diff
@@ -0,0 +1,161 @@
+        try:
+            self.switch.wait_for_service(timeout.to_sec())
+        except rospy.exceptions.ROSException as err:
+            self.fail(
+                "Could not reach controller switch service. Msg: {}".format(err))
+
+    def test_normal_execution(self):
+        """ Test the basic functionality on a straight line """
+        self.move_to_start()
+        self.move_square()
+        self.assertEqual(self.cart_client.get_result().error_code,
+                         FollowCartesianTrajectoryResult.SUCCESSFUL)
+
+    def test_preemption(self):
+        """ Test whether the preemption mechanism works correctly """
+        self.move_to_start()
+        self.move_square(wait_for_result=False)
+        time.sleep(3)  # stop somewhere during execution
+        self.cart_client.cancel_goal()
+        time.sleep(1)
+        self.assertEqual(self.cart_client.get_state(),
+                         GoalStatus.PREEMPTED)
+
+    def test_invalid_goals(self):
+        """ Test whether invalid goals are rejected correctly """
+
+        self.move_to_start()
+
+        # Time from start is not strictly increasing.
+        # This also covers points without time_from_start (they are all implicitly 0).
+        p1 = CartesianTrajectoryPoint()
+        p1.time_from_start = rospy.Duration(1.0)
+        p2 = CartesianTrajectoryPoint()
+        p2.time_from_start = rospy.Duration(0.9999)
+        goal = FollowCartesianTrajectoryGoal()
+        goal.trajectory.points.append(p1)
+        goal.trajectory.points.append(p2)
+        self.cart_client.send_goal(goal)
+        self.cart_client.wait_for_result()
+        self.assertEqual(self.cart_client.get_result().error_code,
+                         FollowCartesianTrajectoryResult.INVALID_GOAL)
+
+    def move_to_start(self):
+        srv = SwitchControllerRequest()
+        srv.stop_controllers = ['cartesian_trajectory_controller']
+        srv.start_controllers = ['joint_trajectory_controller']
+        srv.strictness = SwitchControllerRequest.BEST_EFFORT
+        self.switch(srv)
+
+        q_start = [0, -2.0, 2.26, -0.2513274122872, 1.57, 0.0]  # From base to tip
+        start_joint_state = FollowJointTrajectoryGoal()
+        start_joint_state.trajectory.joint_names = [
+            'joint1',
+            'joint2',
+            'joint3',
+            'joint4',
+            'joint5',
+            'joint6']
+
+        start_joint_state.trajectory.points = [
+            JointTrajectoryPoint(positions=q_start, time_from_start=rospy.Duration(3.0))]
+        self.set_joints.send_goal(
+            start_joint_state)
+        self.set_joints.wait_for_result()
+
+        srv = SwitchControllerRequest()
+        srv.stop_controllers = ['joint_trajectory_controller']
+        srv.start_controllers = ['cartesian_trajectory_controller']
+        srv.strictness = SwitchControllerRequest.BEST_EFFORT
+        self.switch(srv)
```

Comment:
Wouldn't it make more sense to move this block to the function that uses the Cartesian controller? Assuming that this function is called before `move_square()` makes sense, but is not necessarily directly obvious for future developers.

Reply:
Yes, you might be right. It's now a separate call every time we use Cartesian control, see 2c384d3